src/algebra/algebra/basic.lean

```diff
@@ -1359,6 +1309,70 @@ instance int_algebra_subsingleton : subsingleton (algebra ℤ R) :=
 
 end int
 
+namespace no_zero_smul_divisors
+
+variables {R A : Type*}
+
+open algebra
+
+section ring
+
+variables [comm_ring R]
+
+/-- If `algebra_map R A` is injective and `A` has no zero divisors,
+`R`-multiples in `A` are zero only if one of the factors is zero.
+
+Cannot be an instance because there is no `injective (algebra_map R A)` typeclass.
+-/
+lemma of_algebra_map_injective
+  [semiring A] [algebra R A] [no_zero_divisors A]
+  (h : function.injective (algebra_map R A)) : no_zero_smul_divisors R A :=
+⟨λ c x hcx, (mul_eq_zero.mp ((smul_def c x).symm.trans hcx)).imp_left
+  ((injective_iff_map_eq_zero (algebra_map R A)).mp h _)⟩
+
+variables (R A)
+lemma algebra_map_injective [ring A] [nontrivial A]
+  [algebra R A] [no_zero_smul_divisors R A] :
+  function.injective (algebra_map R A) :=
+suffices function.injective (λ (c : R), c • (1 : A)),
+by { convert this, ext, rw [algebra.smul_def, mul_one] },
+smul_left_injective R one_ne_zero
+
+variables {R A}
+lemma iff_algebra_map_injective [ring A] [is_domain A] [algebra R A] :
+  no_zero_smul_divisors R A ↔ function.injective (algebra_map R A) :=
+⟨@@no_zero_smul_divisors.algebra_map_injective R A _ _ _ _,
+ no_zero_smul_divisors.of_algebra_map_injective⟩
+
+end ring
+
+section char_zero_domain
+
+variables [ring R] [is_domain R] [char_zero R]
```

Comment:
```suggestion
variables [ring R] [no_zero_divisors R] [char_zero R]
```
Also you can use `[semiring R]` for `char_zero.no_zero_smul_divisors_nat` after #14414 and #14415 are merged.

Reply:
These are now merged

Reply:
I merged master and generalised as suggested.